Stabilized webgl_gpgpu_water simulation

yomboprime opened the pull request:
Fixed increasing water height level.

Fixes #14800

yomboprime committed:
Fixes #14800

yomboprime committed:
Removing test parameters

yomboprime committed:
Changed water reset feedback to sample 1 point

WestLangley commented on examples/webgl_gpgpu_water.html:
```diff
@@ -122,6 +126,67 @@
 
 		</script>
+		
+		<!-- This is a 'compute shader' to calculate the current volume of water at the center -->
```
You mean current _level_ of the water.

There are multiple occurrences of the word 'volume' in this PR.

yomboprime replied:
Fixed.

WestLangley commented on examples/webgl_gpgpu_water.html:
```diff
@@ -550,6 +657,30 @@
+				// Apply gradually height compensation to reset water level
+				if ( heightCompensation !== 0 ) {
+				
+					var decremHeight = Math.min( 0.2, Math.abs( heightCompensation ) ) * Math.sign( heightCompensation );
```
What is the reason for 0.2 ?

yomboprime replied:
It is just the fixed amount of drain every frame

yomboprime replied:
The exact amount is computed (`heightCompensation`), but applied progressively in each frame (`decremHeight`) after the reading so there is no visual bump in the water level.

You can test by editing `var decremHeight = heightCompensation; heightCompensation = 0;` and will see the visual bump each 2 seconds (or longer period if you want)

WestLangley replied:
As I said, why not decrement by height / 1200 progressively every frame for 1200 frames, then recompute hight and repeat.

Why does your formula contain `Math.min( 0.2, ... )` ?

yomboprime replied:
Now I understand, it makes sense. I've done a commit.

WestLangley replied:
I am not going to be able to invest any more time here...

Perhaps you will agree this is better.

```javascript
if ( ++ numFrames > 120 ) {

	numFrames = 0;

	// Apply gradually height compensation to reset water level

	heightmapVariable.material.uniforms.heightCompensation.value = - readWaterLevel() / 120;

}
```

yomboprime replied:
I'm sorry, but that would apply only one every 120 frames so it wouldn't be gradual. Also, the drain amount would be 1/120 times the needed one.

yomboprime committed:
Substitute water level for volume

yomboprime committed:
Distributed evenly the drain across time.

yomboprime committed:
Compute uniform value only once.

mrdoob commented:
I have the feeling that porting the original water effect to glsl may result in simpler code and a more satisfying and stable simulation.

https://web.archive.org/web/20080618181901/http://freespace.virgin.net/hugo.elias/graphics/x_water.htm

I ported it some years ago to Canvas2D:

https://mrdoob.com/lab/javascript/effects/water/02/

@yomboprime would you like to give it a go?

yomboprime committed:
Changed water algorithm and added tennis balls

yomboprime commented:
I've done it and it looks nice so I've deleted the old algorithm. The waves propagates differently (perhaps just faster) I think I could vary the waves speed by multiplying with a factor in the equations.

I've also added some tennis balls with dynamics because I wanted to keep the read back of textures, just because it is instructive for the reader.

mrdoob commented:
Thanks!
